CIP-0133/README.md

```diff
@@ -0,0 +1,116 @@
+- **ATALA** - a decentralized identification mechanism. One of the properties it can provide is anonymity: users can selectively disclose attributes of their certificate or prove statements without disclosing their identity. Up to date, the most efficient solutions for doing that use pairing-based zero-knowledge protocols.
+
+In conclusion, integrating multi-scalar multiplication as a core built-in within Plutus is not only essential for enhancing cryptographic capabilities, but also for optimizing on-chain computations. The current approach of naive manual implementation in Plutus is inefficient for large-scale MSMs. Incorporating this function directly will streamline these operations, reduce transaction costs, and maintain the integrity of existing tools, thereby significantly advancing the Plutus ecosystem's functionality and user experience.
+
+## Specification
+The MSM for BLS12-381 is implemented in [blst](https://github.com/supranational/blst/blob/e99f7db0db413e2efefcfd077a4e335766f39c27/bindings/blst.h#L241) library, which is already a dependency for the [cardano-base](https://github.com/IntersectMBO/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/). This library provides efficient and formally verified implementation of operations over the BLS12-381 elliptic curve. It has been used for implementing [CIP-0381](https://cips.cardano.org/cip/CIP-0381). Basically, we would like to expose several additional functions from this library in Plutus API.
+
+### Function definition
+We propose to define two new Plutus functions **bls12_381_G1_multiScalarMul** and **bls12_381_G2_multiScalarMul** as follows:
+
+```hs
+bls12_381_G1_multiScalarMul :: [Integer] -> [bls12_381_G1_element] -> bls12_381_G1_element
+bls12_381_G2_multiScalarMul :: [Integer] -> [bls12_381_G2_element] -> bls12_381_G2_element
+```
+
+The types **bls12_381_G1_element** and **bls12_381_G2_element** are already introduced by [CIP-0381](https://cips.cardano.org/cip/CIP-0381). Given two arrays of scalars and group elements the functions compute multi-scalar multiplication for the corresponding subgroup. The arrays of scalars and group elements must be non-empty and have equal size.
+These new functions naturally append a set of operations over BLS12-381 defined by [CIP-0381](https://cips.cardano.org/cip/CIP-0381).
+
+### Cost model
+The computational impact of multi-scalar multiplication is complexified by it having dynamic-size arguments. Preliminary [benchmarks](https://github.com/dkaidalov/bench-blst-msm/) show that the computational complexity grows linearly with the size of the MSM. This should be reflected in the costing function.
+It should also be taken into account that the efficiency of the MSM algorithm may vary [depending on the blst setup](https://github.com/supranational/blst/blob/master/src/multi_scalar.c#L61).
+
```

Comment:
good point, to be honest I didn't investigate blst source code very deeply. 
The low-level C API for MSM is fairly complicated and indeed not well-documented.
What I would suggest is to stick to Pippenger algorithm without table pre-computation, as it is done in [this](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L133) Rust API.

In fact our blst benchmarks use these Rust bindings without table pre-computation. I'm not even sure, would that be feasible to pass around multi-megabyte table in Plutus smart contract?
Also we should keep in mind that our use cases for MSM would be fairly limited. We wouldn't compute more than several hundred points due to script size constraints. So maybe that would be redundant. 
@perturbing any thoughts?

Reply:
Isn't it the case that the rust bindings implement a parallelized + divide-and-Conquer technique for <32 points, since [this](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L61) comment says that the overhead of doing it via a table (which _is_ the Pippenger algo that the C lib implements) is more overhead and slower? @djetchev can you confirm my understanding?

I think we just need to do the same, write the Haskell bindings so that for <32 points it defaults to a naive but optimized MSM, and for larger it uses the C lib. That said, maybe you can experiment with this number 32 when benchmarking @kwxm ? If the costing model is more favorable around another number, why not do that ;)

Reply:
The MSM implementation is quite complicated in blst, but if I understood the code correctly, [that](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L60) comment says about different thing. It compares two-step msm procedure and Pippenger algorithm, which seems to be somewhat different algorithms.

Mentioned two-step MSM with pre-computation is implemented by [`ptype##s_precompute_wbits`](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L122) and [`ptype##s_mult_wbits`](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L179).
It allows to pre-compute a table for a fixed set of points and then reuse it for MSM with different sets of scalars, in which case it is faster than Pippenger. On the other hand, the [Pippenger procedure](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L397) is a different algo which doesn't allow passing the pre-computed table. 
Even more confusing is that the exposed library call for [Pippenger](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L397) internally makes a [selection](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L408)between the two-step MSM (without table re-usability) and pure Pippenger.

The difference between two-step MSM with precompute and Pippenger can be clearly observed in [C++ bindings](https://github.com/supranational/blst/blob/master/bindings/blst.hpp#L418). It provides an interface to use both methods. On the other hand Rust bindings defaults to use only [Pippenger](https://github.com/supranational/blst/blob/415d4f0e2347a794091836a3065206edfd9c72f3/src/multi_scalar.c#L397) procedure without the ability to pass a pre-computed table.

On top of all of that Rust bindings implement its own optimizations with parallelization. In case of more than 2 CPUs and <32 points it defaults to naive parallelized [implementation](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L165). This I believe the cause of time peculiarity at the bound of 32 point. And given how drastically time is improved after 32, I'm not sure this optimization is worthwhile (but it probably depends on how many CPUs are present).

All in all, for Haskell implementation I would suggest to stick to the [implementation](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L133) provided by Rust bindings and just reimplement it in Haskell but with careful consideration whether we need this <32 points naive optimization.

P.S: this is my current understanding after quick code review, but I still may miss something

Reply:
> I'm not even sure, would that be feasible to pass around multi-megabyte table in Plutus smart contract?

I didn't realise that the table would be that big.  I think that would be a problem from the pioint of view of memory costing.  When we call a primitve function in Plutus Core there's a charge for the amount of memory required by the result of the function (we can't really account for transient memory usage during the execution of a foreign function, but we assume that any memory used will be reclaimed in some way when the function returns (except for the return value, which may persist in the Haskell heap)).  The current on-chain limit for total memory usage is 14,000,000 eight-byte words, or 112,000,000 bytes, so a multi-megabyte table could really push memory usage towards the limit.  We do have to be careful about things like this to prevent a malicious script from allocating huge amounts of memory since that might cause problems for the node.

Do you have examples where multiple uses of multiscalar multiplication in the same group are actually necessary?  If that doesn't happen in real programs then we could just include the table creation step in the MSM function. If we did that then we wouldn't have to account for the size of the table since it would become available for garbage collection when the function returns. Of course the gains from pre-computation would disappear then as well.

Reply:
> This I believe the cause of time peculiarity at the bound of 32 point. And given how drastically time is improved after 32, I'm not sure this optimization is worthwhile (but it probably depends on how many CPUs are present).

We'd probably just charge for the worst case anyway, so any efficiency improvments for small cases might not be visible in the costs the user actually has to pay.  We'll have to see how the Plutus Core  implementation behaves in practice before we decide hoow the costing will actually work though.

We also don't take any account of parallelism since we don't know how many cores nodes on the network have (so we'd probably base costs on the performance on a single core).  In fact, I'm not sure if we're even allowed to use code that uses multiple cores on the chain. I'll try to check that.

Reply:
> What I would suggest is to stick to Pippenger algorithm without table pre-computation, as it is done in [this](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L133) Rust API.

Yes, that sounds like the simplest thing to do.  We can at least do that initially and seehow it goes.

Reply:
In case of single-threaded implementation you can take a look at [this](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger-no_std.rs#L78) Rust bindings to see how to use C interface.

Reply:
@kwxm could you please resolve the conversations if there are no more questions